docs/deployment-examples/custom-scripts/script-modules.md

```diff
@@ -82,14 +67,144 @@ Make sure to select a **Role**, an **Environment** and to put a **Step Name**
+in scope for your script.
+
+Script Module:
+```powershell
+function Say-Hello($name) {
+    Write-output "Hello $name. Welcome to Octopus!"
+}
+```
+
+Script Step:
+```powershell
+Say-Hello -name "George"
+```
+
+### Gotcha
+If you process has 4 steps (i.e Stop IIS, Backup, Deploy, Start IIS) the entire module will get loaded once at the start of each step. Because of this we encourage users to avoid putting code outside functions on your Script Modules. This way the code will only get executed when the function is properly called from a PowerShell Script step.
+
+In the example Script Module below, the first line which attempts to stop the service "ImportantService" will run for every PowerShell-Script-based step on your deployment. The first time it might succeed (stopping the service), but the subsequent tries will most likely make the overall deployment fail.
+
+```powershell
+# bad example - dont copy
+Stop-Service -Name ImportantService
+ 
+function Say-Hello($name) {
+    Write-output "Hello $name. Welcome to Octopus!"
+}
+```
+
+## Bash Script Modules{#ScriptModules-Bash}
+
+Bash Script Modules are written as a `.sh` file next to your script. Import them
+via `source MyScriptModule.sh`, where `MyScriptModule` is the name of your Script
+Module with invalid characters removed. The help text above the Script Module body
+will show the filename that will be created.
+
+Given a Bash Script Module called `Bash Script Module`:
+```bash
+say_hello() {
+    echo "Hello $1. Welcome to Octopus!"
+}
+```
+
+Call it from your Script Step with:
+```bash
+source BashScriptModule.sh
+say_hello George
+```
+
+## C# Script Modules
+
+C# Script Modules are written as a `.csx` file next to your script. Import them
+via `#load "MyScriptModule.csx"`, where `MyScriptModule` is the name of your Script
+Module with invalid characters removed. The help text above the Script Module body
+will show the filename that will be created.
+
+The `#load` statement must be at the top of your script.
+
+Given a C# Script Module called `CSharp Script Module`:
+```csharp
+public void SayHello(string name) {
+    Console.WriteLine("Hello " + name + ". Welcome to Octopus!");
+}
+```
+
+Call it from your Script Step with:
+```csharp
+#load "CSharpScriptModule.csx"
+SayHello("George");
+```
+
+## F# Script Modules
+
+F# Script Modules are written as an `.fsx` file next to your script. Import them
+via `#load "MyScriptModule.fsx"`, where `MyScriptModule` is the name of your Script
+Module with invalid characters removed. The help text above the Script Module body
+will show the filename that will be created.
+
+F# script modules need to declare a module name at the top of the Script Module body with `module MyModule`.
+
+Once the Script Module file is loaded, you can either use the functions from the Script Module by prefixing the functions with the module name, or by "opening" the Script Module with `open MyModule`:
+
+```fsharp
+// call the function in the module directly
```

Comment:
Maybe it's worthwhile to add the #load here again to really drill the point home that you have to include it.